docs/design-documents/platform/clocks/clocks.md

```diff
@@ -0,0 +1,381 @@
+
+Note that the `TickerDataClock` is a non-standard Chrono clock, in that its `now` method is non-static. The Chrono type-checking will not detect that `TickerDataClock::time_point`s for different tickers are not interchangable. It is generally preferable to use the dedicated `HighResClock` or `LowPowerClock`, but `TickerDataClock` is used by some generic code, to avoid template bloat from separate code for each Clock. 
+
+The base classes `TimerBase`, `TickerBase` and `TimeoutBase` use `TickerDataClock` to provide the core of the high-level implementation, and these are then used to provide the separate strongly-typed high-resolution and low-power classes above.
+
+Systems wishing to provide more HAL clocks beyond the basic two can use the same framework to give that clock its own custom `Clock/Timer/Ticker/Timeout` set of classes, using the base classes.
+
+### The real-time clock
+
+The real-time clock, if available, provides calendar date and time information, at one-second resolution. The RTC can continue to run while the CPU does not.
+
+The C APIs in `rtc_api.h` read and write the time as a C `time_t`.
+
+The class `RealTimeClock` provides a wrapper that acts as a Chrono clock, so that `RealTimeClock::time_point`can be used to express absolute real time.
+
+The methods `to_time_t` and `from_time_t` convert to and from C `time_t`.
+
+At present there are no other APIs using `RealTimeClock`, and no real-time-based wake or alarm facility, so utility is limited, but the class has been created for framework symmetry with the other clocks.
+
+### Comparison to standard C++ Chrono clocks
+
+Mbed OS does not provide the standard C++11 Chrono Clocks, as it is not possible to provide consistent behaviour across the various toolchains and libraries supported. Here's a suggested mapping from the 3 standard C++ Clocks to Mbed OS.
+
+**std::steady_clock**
+
+`std::steady_clock` represents a clock for which values of `time_point` never decrease as physical time advances and for which values of `time_point` advance at a steady rate relative to real time.
+
+`Kernel::Clock` can be used in place of `std::steady_clock` for general use.
+
+For slightly higher timestamp resolution you can use `LowPowerClock`, but if attempting to schedule time events, using RTOS APIs with `Kernel::Clock` will often work better due to the RTOS's deep sleep wake latency compensation.
+
+If high resolution is required, you can use `SteadyHighResClock`, but that will prevent deep sleep from occurring at all.
+
+**std::high_resolution_clock**
+
+`std::high_resolution_clock` represents the clock with the shortest tick period.
+
+`HighResClock` is the best choice for a high-resolution clock, but it will not run during deep sleep. You can use `HighResClock::lock` to keep it running, or otherwise ensure that deep sleep is locked.
```

Comment:
I get your point, but the motivation is to make it generic - it means that `HighResClock` is a C++ _Lockable_. So it could be used with `scoped_lock`/`lock_guard`/`ScopedLock` or similar.

    my_func()
    {
        HighResClock clock; // dummy object - ScopedLock needs an object to work on, regardless of our methods being static
        ScopedLock<XXXClock> lock(clock);
        do stuff();
        if (error) {
             return;
        }
        do more stuff();
    }

The scoped lock automatically cleans up for you on any return.

A Lockable can be positive (locking in or on) or negative (locking out or off), and we do already have "positive"-named ones like `CriticalSectionLock` and `ScopedRamExecutionLock`. The common thing is that you're locking the temporary state (deep sleep off, IRQs off, mutual exclusion), and then returning to the default state (deep sleep on, IRQs on, no mutual exclusion).

So here you're locking on "clock is running", which is the temporary state, then returning to the default "clock is not running" state.

Arguably it's `DeepSleepLock` which should be named more positively - `ShallowSleepLock` or something, indicating it must maintain that power state or higher, rather than indicating the power state you don't want to go to.

Reply:
I understand! I did not know about C++ Lockable and you're the expert! :)

For my personal knowledge, is that what we're talking about? http://www.cplusplus.com/reference/concept/Lockable/

Reply:
Yes, that's what I meant, but the reference corrects me; it's only _BasicLockable_, due to absence of the `try_lock`.

Reply:
I might be missing something, but:

I understand you can lock a mutex to get exclusive access to a specific resource or section of code and prevent other threads from running this part until the mutex is unlocked. (example: http://cpp.sh/6wcdi)

But for the XXXClock the `lock` doesn't directly refer the the Clock but instead it refers to the DeepSleep state that you want to avoid. 

But the XXXClock is not really a shared resource, is it?

Maybe this part is what I'm missing:

> You need to be Lockable rather than just BasicLockable if locking more than 1 thing for the multi-lock anti-deadlock algorithm to work.

Reply:
A "lock" can lock anything.   It doesn't need to be a mutex (mutual exclusion). A mutex is only one of the types of things that can be locked.

You can also lock a door, the running of a clock, preventing thread switches, preventing IRQs, staying awake, or keeping the power state above at least some level.

"Locking" the clock means you're using it. Just like "locking" the mutex means you're using it. But only for the mutex does "using it" mean "mutual exclusion", because that's its specific meaning. For the clock "using it" just means "keep it running".

So that example above `scoped_lock<Mutex, HighResClock> lock(m, clock);` means "lock these two objects for the lifetime of this lock object".  (In C++17, you could just write it more simply as `scoped_lock lock(m, clock)`).

Which in this case would ensure mutual exclusion on the mutex - no-one else locking m can run at the same time - and ensuring running of the clock. Two different things, but both handled by the generic "lock" concept.

And "lock the running of the clock" is an abstraction. For LowPowerClock, it's a no-op. For HighResClock it means "lock deep sleep".

That multi-lock includes an ordering algorithm to prevent deadlock. Conceivably someone taking two mutexes could cause deadlock if they kept holding mutex A while waiting for B, but someone else was holding B while waiting for A. It will internally release the others while waiting for one.

In this clock+mutex case, that wouldn't be a deadlock issue, but it would automatically make sure you didn't keep the high-res clock running while waiting for the mutex, so a slight power saving.

Of course, you can do that manually by just writing

     mutex.lock();
     HighResClock::lock();

in that order because you "know"  `mutex.lock()` is a blocking operation but `HighResClock::lock()` isn't, but that wouldn't necessarily be safe for a pair of mutexes.

Reply:
> But the XXXClock is not really a shared resource, is it?

Oh, yes, it is, but there's no mutual exclusion requirement.  It can be running, and multiple people can be reading its `time_point`s to monitor the time base.

There is internal critical section locking inside `ticker_read` and other APIs to deal with mutual exclusion for the brief duration of API calls, but that's not visible to users.

In API we only need the higher-level "lock it running" concept, which is thus a reference count about how many people are currently relying on it running.

Reply:
Thanks for the detailed explanation ! :) it makes perfect sense now. Being a non-native English speaker (French), `lock` meant something like "prevent others from using it" or "prevent this from running" instead of "I'm using it".

> And "lock the running of the clock" is an abstraction. For LowPowerClock, it's a no-op. For HighResClock it means "lock deep sleep".

As you said earlier, in this case the `DeepSleepLock` is not well named ;)

Reply:
I think this generalisation of "lock" in computer science may be a relatively recent thing. I'd agree the meaning is a little forced in some cases. :)